src/impl_/pymethods.rs

```diff
@@ -585,6 +585,11 @@ pub fn inspect_type<T>(t: T) -> (T, Extractor<T>) {
     (t, Extractor::new())
 }
 
+#[allow(clippy::type_complexity)]
+pub fn inspect_fn<A, T>(f: fn(A) -> PyResult<T>) -> (fn(A) -> PyResult<T>, Extractor<A>) {
+    (f, Extractor::new())
+}
```

Comment:
Smart! In #3847 (which I'm slowing making progress on) I recently found that it's also possible to take the `Extractor` as an argument by reference:

```suggestion
#[allow(clippy::type_complexity)]
pub fn inspect_fn<A, T>(f: fn(A) -> PyResult<T>, _: &Extractor<A>) -> fn(A) -> PyResult<T> {
    f
}
```

The nice thing about this is that the dance in the macro code gets simpler. Instead of:

```rust
let extractor;
let (thing, e) = inspect_fn(thing);
extractor = e;
```

can just write:

```rust
let extractor = Extractor::new();
let thing = inspect_fn(thing, &e);
```

So far this seems to be working for me for type inference just as well as the original form and seems to be easier to fit into macro code paths, I wonder if it will also simplify the return type and make clippy happier.

Reply:
Interesting! That does look nicer and easier to work with. In this particular case it does not make too much of a difference, because I did not need that empty binding, but it does not hurt either (and it makes clippy indeed happier)